src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -314,29 +329,41 @@ protected override void ApplyParseOptionsChanged(ProjectId projectId, ParseOptio
                 throw new ArgumentNullException(nameof(options));
             }
 
-            var parseOptionsService = CurrentSolution.GetProject(projectId).LanguageServices.GetService<IParseOptionsService>();
-            Contract.ThrowIfNull(parseOptionsService, nameof(parseOptionsService));
+            var storage = ProjectPropertyStorage.Get(TryGetDTEProject(projectId), DeferredState.ServiceProvider);
 
-            string newVersion = parseOptionsService.GetLanguageVersion(options);
+            switch (options)
+            {
+                case CSharpCompilationOptions csharpOptions:
+                    storage.SetProperty("AllowUnsafeBlocks", nameof(ProjectConfigurationProperties3.AllowUnsafeBlocks),
+                        csharpOptions.AllowUnsafe.ToString().ToLowerInvariant());
+                    break;
+                case VisualBasicCompilationOptions _:
+                    throw new InvalidOperationException(ServicesVSResources.This_workspace_does_not_support_updating_Visual_Basic_compilation_options);
+            }
+        }
 
-            GetProjectData(projectId, out var hostProject, out var hierarchy, out var project);
-            foreach (string configurationName in (object[])project.ConfigurationManager.ConfigurationRowNames)
+        protected override void ApplyParseOptionsChanged(ProjectId projectId, ParseOptions options)
+        {
+            if (projectId == null)
             {
-                switch (hostProject.Language)
-                {
-                    case LanguageNames.CSharp:
-                        var csharpProperties = (VSLangProj80.CSharpProjectConfigurationProperties3)project.ConfigurationManager
-                            .ConfigurationRow(configurationName).Item(1).Object;
-
-                        if (newVersion != csharpProperties.LanguageVersion)
-                        {
-                            csharpProperties.LanguageVersion = newVersion;
-                        }
-                        break;
+                throw new ArgumentNullException(nameof(projectId));
+            }
```

Comment:
Nit: there should be a line break after this closing brace.

Reply:
It's easy to miss in the diff, but there is one :smile:

Reply:
did this function change?  it's hard to tell in the diff.